src/main/scala/analysis/DSAUtility.scala

```diff
@@ -0,0 +1,837 @@
+          case _ => m
+      case MemoryAssign(mem, index: Expr, value: Expr, endian, size: Int, label) =>
+        visitStackAccess(pos, index, size, m)
+      case _ => m
+
+  }
+
+  private def createStackMapping(label: String, offset: BigInt, m:  Map[BigInt, DSN], byteSize: Int) :  Map[BigInt, DSN]=
+    if m.contains(offset) then
+      assert(!m(offset).cells(0).growSize(byteSize))
+      m
+    else
+      val node = DSN(Some(this), byteSize)
+      node.allocationRegions.add(StackLocation(label, proc, byteSize))
+      node.flags.stack = true
+      node.addCell(0, byteSize)
+      m + (offset -> node)
+
+
+  private val swappedOffsets = globalOffsets.map(_.swap)
+
+  // creates the globals from the symbol tables
+  val globalMapping: mutable.Map[AddressRange, Field] = mutable.Map[AddressRange, Field]()
+  globals.foreach(
+    global =>
+      val node = DSN(Some(this), global.size)
+      node.allocationRegions.add(DataLocation(global.name, global.address, global.size/8))
+      node.flags.global = true
+      node.flags.incomplete = true
+      globalMapping.update(AddressRange(global.address, global.address + global.size/8), Field(node, 0))
+  )
+
+  // creates a global for each relocation entry in the symbol table
+  // the global corresponding to the relocated address points to the global corresponding to the original address
+  globals.foreach(
+    global =>
+      var address = global.address
+      breakable {
+        while swappedOffsets.contains(address) do
+          val relocatedAddress = swappedOffsets(address)
+          if relocatedAddress == address then
+            break
+
+          var field: BigInt = 0
+          val node: DSN = isGlobal(relocatedAddress) match
+            case Some(value) =>
+              field = relocatedAddress - value._1._1
+              val node = value._2._1
+              node.addCell(field, 8)
+              node
+
+            case None =>
+              val node = DSN(Some(this))
+              node.allocationRegions.add(DataLocation(s"Relocated_$relocatedAddress", relocatedAddress, 8))
+              node.flags.global = true
+              node.flags.incomplete = true
+              globalMapping.update(AddressRange(relocatedAddress, relocatedAddress + 8), Field(node, 0))
+              node
+
+          pointTo.update(node.cells(field), Slice(isGlobal(address).get._2._1.cells(0), 0))
+          address = relocatedAddress
+      }
+  )
+
+  externalFunctions.foreach(
+    external =>
+      val node = DSN(Some(this))
+      node.allocationRegions.add(DataLocation(external.name, external.offset, 0))
+      node.flags.global = true
+      node.flags.incomplete = true
+      globalMapping.update(AddressRange(external.offset, external.offset), Field(node, 0))
+  )
+
+
+
+  // determine if an address is a global and return the corresponding global if it is.
+  def isGlobal(address: BigInt): Option[(AddressRange, Field)] =
+    var global: Option[(AddressRange, Field)]  = None
```

Comment:
This is another case that would benefit from using a case class instead of a tuple